batch2/batch/front_end/front_end.py

```diff
@@ -137,139 +149,231 @@ def create_job(app, jobs_builder, batch_id, job_spec):  # pylint: disable=R0912
-    if not validator.validate(parameters):
-        raise web.HTTPBadRequest(reason='invalid request: {}'.format(validator.errors))
-
-    batch = await Batch.create_batch(
-        request.app,
-        attributes=parameters.get('attributes'),
-        callback=parameters.get('callback'),
-        userdata=userdata,
-        n_jobs=parameters['n_jobs'])
-    if batch is None:
-        raise web.HTTPBadRequest(reason=f'creation of batch in db failed')
+    app = request.app
+    db = app['db']
 
-    log.info(f'took {round(time.time() - start, 3)} seconds to initialize batch {batch.id} in db')
-    return web.json_response(await batch.to_dict(include_jobs=False))
+    batch_spec = await request.json()
 
+    validator = cerberus.Validator(schemas.batch_schema)
+    if not validator.validate(batch_spec):
+        raise web.HTTPBadRequest(reason=f'invalid request: {validator.errors}')
 
-async def _get_batch(app, batch_id, user, limit=None, offset=None):
-    batch = await Batch.from_db(app, batch_id, user)
-    if not batch:
+    user = userdata['username']
+    attributes = batch_spec.get('attributes')
+    async with db.pool.acquire() as conn:
+        await conn.begin()
+        async with conn.cursor() as cursor:
+            await cursor.execute(
+                '''
+INSERT INTO batch (userdata, user, attributes, callback, n_jobs)
+VALUES (%s, %s, %s, %s, %s);
+''',
+                (json.dumps(userdata), user, json.dumps(attributes),
+                 batch_spec.get('callback'), batch_spec['n_jobs']))
+            id = cursor.lastrowid
+
+        if attributes:
+            async with conn.cursor() as cursor:
+                await cursor.executemany(
+                    '''
+INSERT INTO `batch-attributes` (batch_id, `key`, `value`)
+VALUES (%s, %s, %s)
+''',
+                    [(id, k, v) for k, v in attributes.items()])
+        await conn.commit()
+
+    return web.json_response({'id': id})
+
+
+async def _get_batch(app, batch_id, user, include_jobs):
+    db = app['db']
+
+    record = await db.execute_and_fetchone(
+        '''
+SELECT * FROM batch
+WHERE user = %s AND id = %s AND NOT deleted;
+''', (user, batch_id))
+    if not record:
         raise web.HTTPNotFound()
-    return await batch.to_dict(include_jobs=True, limit=limit, offset=offset)
+    return await batch_record_to_dict(db, record, include_jobs=include_jobs)
 
 
 async def _cancel_batch(app, batch_id, user):
-    batch = await Batch.from_db(app, batch_id, user)
-    if not batch:
+    db = app['db']
+
+    record = await db.execute_and_fetchone(
+        '''
+SELECT closed FROM batch
+WHERE user = %s AND id = %s AND NOT deleted;
+''',
+        (user, batch_id))
+    if not record:
         raise web.HTTPNotFound()
-    await batch.cancel()
+    if not record['closed']:
+        raise web.HTTPBadRequest(reason='cannot cancel open batch {batch_id}')
+
+    await db.execute_update(
+        'UPDATE batch SET cancelled = closed WHERE id = %s;', (batch_id,))
```

Comment:
Shouldn't cancelled be a boolean here?

Reply:
It is; so is closed.  Only a closed batch should be cancelled, so this is slightly safer than "cancelled = 1".